### PR TITLE
deprecated reflect mode has been replaced with package mode

### PR DESCRIPTION
**Pull request by tulzke**
re-open of: https://github.com/uber-go/mock/pull/198

resolves https://github.com/uber-go/mock/issues/175
resolves https://github.com/uber-go/mock/issues/197
resolves https://github.com/uber-go/mock/issues/128

It is impossible to create an mock for a generic interface via reflect mode, because it is impossible to compile a generic type without instantiation.
This PR replaces the reflect mod for parsing using go/types.

All exists mocks have been regenerated and the tests have been passed. But since this radically changes the behavior of reflect mode, I would be grateful if there are those who want to add additional test cases that I did not provide.

We can also come up with another name instead of import mode.

benefits:
generation mocks for generic interfaces
generation mocks for aliases to interfaces
correct names for method arguments

### CODE REVIEW

**Comment by tulzke** [comment]
@JacobOaks 
What do you think about this problem?
https://github.com/uber-go/mock/pull/198#discussion_r1777459359

**Comment by JacobOaks** [comment]
@tulzke - thanks for re-opening this and apologies again for the git mistakes on the other PR.

I think for now it makes sense to just use the unaliased names until we're able to update `go.mod` to 1.23. From the perspective of type-checking, types and their aliases are interchangeable so there shouldn't be any missing functionality/bugs introduced by just uniformly using the unaliased names for now. This also keeps the generated code consistent between the two versions. Once 1.24 is out and we can upgrade `go.mod` to 1.23, we can definitely then switch over to using the alias names.

I think the changes to do this aren't too bad, we would just have a separate case in `parseType` like so:
```
case *types.Alias:
    return p.parseType(types.Unalias(t))
```
What are your thoughts?

**Comment by tulzke** [comment]
> we would just have a separate case

I think, that alias for us is the same thing as a `*types.Named`. We need only his name and package.

`types.Unalias(t)` returns original type, so in mocks we will see original types. It is not completely correct behavior.

**Comment by tulzke** [comment]
So, if we have agreed on the functionality, can we think about the names? What is the name of the new mode to replace reflect? 
Maybe Package mode? Source mode and Package mode. That sounds good.

**Comment by JacobOaks** [comment]
I don't have a strong opinion on a name. "package mode" or "types mode" is fine.

**Comment by tulzke** [comment]
> I don't have a strong opinion on a name. "package mode" or "types mode" is fine.

I choose package mode. Check last commit changes please: https://github.com/uber-go/mock/pull/207/commits/69210e9badcc1aacba501d26e5c40b3ff79de2d9

If everything is OK, i'm waiting for a merge 🙂

**Comment by tulzke** [comment]
I realized that I had made a mistake. 
I deleted 3 unused flags, and this may break the existing generation methods. Since now these flags have no effect, I think they can be removed from the readme and help, and when using write a WARNING message to the console.

@JacobOaks What do you think about this?

**Comment by JacobOaks** [comment]
Hey @tulzke, for `--exec_only` and `--prog_only`, I think we can do what you're suggesting.

For build flags, what do you think about maintaining backwards compatibility and passing them to `packages.Load` via [`Config`](https://pkg.go.dev/golang.org/x/tools/go/packages#Config) which has a `BuildFlags` field? My understanding is these will get passed to the `go list` command, which accepts normal go build flags.

**Comment by tulzke** [comment]
> Hey @tulzke, for `--exec_only` and `--prog_only`, I think we can do what you're suggesting.
> 
> For build flags, what do you think about maintaining backwards compatibility and passing them to `packages.Load` via [`Config`](https://pkg.go.dev/golang.org/x/tools/go/packages#Config) which has a `BuildFlags` field? My understanding is these will get passed to the `go list` command, which accepts normal go build flags. This will keep things like build tags specified by `--tag` to be respected when deciding which Go files to load, for example.

I started developing this feature. I decided to see how it is implemented now. And there was no limit to my surprise. The flag is declared, but not used. 😁
https://github.com/search?q=repo%3Auber-go%2Fmock%20build_flags&type=code
https://github.com/search?q=repo%3Agolang%2Fmock%20build_flags&type=code

**Comment by tulzke** [comment]
@JacobOaks 
I returned support for `build_flags` and added additional test case for it

File: README.md [hunk 1]
```diff
@@ -98,8 +97,6 @@ It supports the following flags:
   `foo=bar/baz.go`, where `bar/baz.go` is the source file and `foo` is the
   package name of that file used by the -source file.
 
-- `-build_flags`: (reflect mode only) Flags passed verbatim to `go build`.
-
```
Comment (JacobOaks) [inline README.md:102]: Should we keep these lines now since we are keeping the flag?

Comment (tulzke) [reply README.md]: You're right, i forgot to do this

**Review by JacobOaks — APPROVED** [review]
Thanks @tulzke for all your hard work on this PR! LGTM.

**Comment by tulzke** [comment]
Hi @JacobOaks 
I'm glad that PR was merged. So, when is the new release?

File: mockgen/package_mode.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+}
+
+func (p *packageModeParser) loadPackage(packageName string) (*packages.Package, error) {
+	var buildFlagsSet []string
+	if *buildFlags != "" {
+		buildFlagsSet = strings.Split(*buildFlags, " ")
+	}
+
+	cfg := &packages.Config{
+		Mode:       packages.NeedDeps | packages.NeedImports | packages.NeedTypes | packages.NeedTypesInfo | packages.NeedEmbedFiles,
```
Comment (johanneswuerbach) [inline mockgen/package_mode.go:49]: @tulzke / @JacobOaks any particular reason that `NeedDeps` was used here? I was looking into mock generation performance and this seems to cause a significant slow down (as expected) as all dependencies are parsed.

With the option removed no tests are breaking and for us its working without, but maybe you had a case in mind that needs `NeedDeps` https://github.com/uber-go/mock/pull/230

**Pull request merged**